File: typeScript/challenges/challenge22.ts
export function generateGiftSets (gifts: string[]): string[][] {
  const result: string[][] = []

  function backtrack (start: number, path: string[]): void {
    // Add the current combination to the result
    if (path.length > 0) {
      result.push([...path])
    }

    for (let i = start; i < gifts.length; i++) {
      // Include gifts[i] in the current combination
      path.push(gifts[i])
      // Move on to the next item
      backtrack(i + 1, path)
      // Backtrack and remove the last item added
      path.pop()
    }
  }

  backtrack(0, [])
  // sorted
  return result.sort((a, b) => a.length - b.length)
}

File: python/challenge22.py
def generate_gift_sets(gifts):
    def backtrack(start, current):
        if len(current) > 0:
            result.append(current[:])  # Agregar la combinación actual
        for i in range(start, len(gifts)):
            current.append(gifts[i])  # Agregar el siguiente regalo
            backtrack(i + 1, current)  # Continuar con la siguiente posición
            current.pop()  # Retroceder para explorar otras combinaciones

    result = []
    backtrack(0, [])
    result.sort(key=len)
    return result








print(generate_gift_sets(['car', 'doll', 'puzzle']))
# [
#   ['car'],
#   ['doll'],
#   ['puzzle'],
#   ['car', 'doll'],
#   ['car', 'puzzle'],
#   ['doll', 'puzzle'],
#   ['car', 'doll', 'puzzle']
# ])